### PR TITLE
feat(vector sink): compression support for vector v2

### LINKED ISSUE #9558: Allow to encode/decode network traffic between vector client and server with zstd/snappy/brotli
### Current Vector Version

```
0.17.0
```
### Use-cases
Currently, we are sending traffic from several dozen clients to a single vector server, at the moment we are approaching 3 Gbps on the input. It would be great if we could start compressing outgoing traffic on the client and automatically decompress it on the server to reduce the load on the network.

### Attempted Solutions
No solution

### Proposal
Add options for client to encode traffic with desired codec. Auto decode traffic on server.

### PR DESCRIPTION
**Pull request by splitice**
Forewarning: I don't daily develop rust. Compared to this code-base, I'm a beginner. Appologies if I havent done anything the way you need it done.

Fixes #9558

Adds `compression` to the sink. v2 source gets the ability to receive compressed messages if supplied.

I wouldnt mind if someone more experienced than I added a test that confirmed the sink is sending compressed messages.

### CODE REVIEW

**Comment by hhromic** [comment]
Nice PR! We use vector-to-vector communication via gRPC a lot in our use cases, thanks for this!
Did you measure any noticeable performance benefit when using compression?
I know that when using compresssion, one is trading-off CPU usage for network usage so I'm curious of the overall improvement, considering that gRPC data (ProtoBuffer) is already binary and not as compressable as text payloads.

Looking forward to test this when merged! :)

File: website/cue/reference/components/sinks/vector.cue [hunk 1]
```diff
@@ -83,6 +83,12 @@ components: sinks: vector: {
 				examples: ["92.12.333.224:\(_port)"]
 			}
 		}
+		compression: {
+			description: "Enable gRPC compression"
```
Comment (hhromic) [inline website/cue/reference/components/sinks/vector.cue:87]: I would make it clear that the compression method is `gzip` (for now).

Comment (splitice) [reply website/cue/reference/components/sinks/vector.cue]: Done

**Comment by splitice** [comment]
@hhromic I havent done any benchmarking, I'm betting however it would depend greatly on the content and batch size.

What we are pushing is highly compressable, so I'm expecting reasonable compression (at-least 8:1 hopefully)

**Comment by hhromic** [comment]
> @hhromic I havent done any benchmarking, I'm betting however it would depend greatly on the content and batch size.
> 
> What we are pushing is highly compressable, so I'm expecting reasonable compression (at-least 8:1 hopefully)

Yes, is true that its effectiveness highly depends on the content and batch sizing.
Looking forward to test/benchmark on our data as I think we have good chances of compressibility. Thanks again for this!

File: src/sinks/vector/v2/config.rs [hunk 2]
```diff
@@ -76,9 +79,9 @@ impl VectorConfig {
             .clone()
             .map(|uri| uri.uri)
             .unwrap_or_else(|| uri.clone());
-        let healthcheck_client = VectorService::new(client.clone(), healthcheck_uri);
+        let healthcheck_client = VectorService::new(client.clone(), healthcheck_uri, false);
```
Comment (jszwedko) [inline src/sinks/vector/v2/config.rs:82]: Should we use the compression setting here too?

**Review by jszwedko — COMMENTED** [review]
Thanks for this contribution @splitice !

I had hoped we could make the compression settings configurable in the same way compression settings are configurable on other sinks (which allow configuring gzip compression level), but I see that the `tonic` crate lacks the ability to configure this so this seems like the best we can do.

Just one comment about the health check, but otherwise this looks good!

File: src/sinks/vector/v2/config.rs (hunk 2 above)
Comment (splitice) [reply src/sinks/vector/v2/config.rs:82]: Being a small message I can't see any gain. Tonic should send plaintext any message that compressed result is larger than the plaintext (I havent checked) with only a waste in CPU. I think disabled is wiser.

Comment (jszwedko) [reply src/sinks/vector/v2/config.rs:82]: Ah, yeah, that's a good point. 👍

**Comment by splitice** [comment]
@jszwedko done

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**